src/io/output.rs

```diff
@@ -93,18 +108,86 @@ fn get_raw_output(vm: &VirtualMachine, output_base: usize, output_size: usize) -
     raw_output
 }
 
+fn next_or_fail<T, I: Iterator<Item = T>>(output_iter: &mut I, item_name: &str) -> Result<T, SnOsError> {
+    output_iter.next().ok_or(SnOsError::CatchAll(format!("Could not read {item_name} field")))
+}
+
+fn next_as_usize<I: Iterator<Item = Felt252>>(output_iter: &mut I, item_name: &str) -> Result<usize, SnOsError> {
+    output_iter
+        .next()
+        .ok_or(SnOsError::CatchAll(format!("Could not read {item_name} segment size")))?
+        .to_usize()
+        .ok_or(SnOsError::CatchAll(format!("{item_name} segment size is too large")))
+}
+
+fn parse_contract_changes<I: Iterator<Item = Felt252>>(output_iter: &mut I) -> Result<ContractChanges, SnOsError> {
+    let addr = next_or_fail(output_iter, "contract change addr")?;
+    let class_nonce_n_changes = next_or_fail(output_iter, "contract change class_nonce_n_changes")?;
+
+    // unwrap() is safe here, we now the value is non zero.
+    let two_exp_64 = Felt252::from(1u128 << 64).try_into().unwrap();
+    let (class_nonce, n_changes) = class_nonce_n_changes.div_rem(&two_exp_64);
+    let (class_updated, nonce) = class_nonce.div_rem(&two_exp_64);
+
+    let class_hash = if class_updated != Felt252::ZERO {
+        Some(next_or_fail(output_iter, "contract change class_hash")?)
+    } else {
+        None
+    };
+
+    // unwrap() is safe because we know that n_changes fits in 64 bits by definition
+    // of the format above.
```

Comment:
This explanation could be used as the text for an `expect()` call instead of `unwrap()`, perhaps at the expense of making the binary a bit larger.

Reply:
Good idea.